---
title: "R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
---

```{r Installations, message=FALSE, warning=FALSE, include=FALSE}

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install(version = "3.19")
BiocManager::install("GenomicRanges")
BiocManager::install("igvR")
BiocManager::install("AnnotationHub")
BiocManager::install("plyranges")

if (!require("arrow", quietly = TRUE))
  install.packages("arrow")

if (!require("tidyverse", quietly = TRUE))
  install.packages("tidyverse")

if (!require("ggsci", quietly = TRUE))
  install.packages("ggsci")

if (!require("ggalluvial", quietly = TRUE))
  install.packages("ggalluvial")

if (!require("ggedit", quietly = TRUE))
  install.packages("ggedit")  

if (!require("ggdist", quietly = TRUE))
  install.packages("ggdist")

```

```{r Libraries, message=FALSE, warning=FALSE, include=FALSE}

library(arrow)
library(tidyverse)
library(esquisse)
library(ggsci)
library(ggalluvial)
library(ggdist)

library(GenomicRanges)
library(rtracklayer)
library(AnnotationHub)
library(Biostrings)
library(igvR)
library(plyranges)
```

We generate a custom function that will return a continuous range of colours on our chosen palette: Futurama

```{r Palette function, message=FALSE, warning=FALSE, include=FALSE}

futurama_unlimited_palette <- function(input_colour_number = 12, output_colour_number) {
  planet_express <- pal_futurama("planetexpress")(input_colour_number)
  output_colour <- colorRampPalette(planet_express)(output_colour_number)
  return (output_colour)
}

futurama_unlimited_palette(3, 100)

```

We import our file in parquet format:

```{r Data}

data <- arrow::read_parquet(
  file.path(
    'F:',
    'Archive',
    'Trabajo',
    'Profesiones',
    'Bioinformatica',
    'PhD',
    'Assignments',
    'enERVate',
    'results',
    'tables',
    'full_vs_ltr.parquet'
  )
)
```

And try to find the columns with the relevant data:

```{r echo=TRUE}

colnames(data)

```

We generate new columns in the dataframe for LTR candidates chromosome, start and end in order to later generate our GenomicRanges:

```{r}

data <- data %>%
        mutate(chromosome = str_extract(alignment_hit_def,  "(?<=chromosome\\s)(\\d{1,2}|[A-Z])")) %>%
        mutate(ltr_start = as.integer(str_extract(alignment_hit_def, "(?<=\\[)\\d+"))) %>%
        mutate(ltr_end = as.integer(str_extract(alignment_hit_def, "(?<=,)\\d+(?=\\])")))

head(data$ltr_start)

```

We change the sequence column to a Biostring for easier, more powerful manipulation. Let's explore the column:

```{r}

data <- data %>%
        mutate(genbank_seq = map(genbank_seq, ~DNAString(.)),
               hsp_query = map(hsp_query, ~AAString(.)),
               hsp_sbjct = map(hsp_sbjct, ~AAString(.)))

head(data$hsp_sbjct)
```

We also generate new columns with Bioproject and Biosample:

```{r}

data <- data %>%
        mutate(bioproject = map_chr(genbank_dbxrefs, ~str_extract(.[1], "(?<=BioProject:)\\w+")),
               biosample = map_chr(genbank_dbxrefs, ~str_extract(.[2], "(?<=BioSample:)\\w+")))

head(data$biosample)
```

Let's also take a look at the distribution of bitscores across all species. We'll convert all non-retroviral probes into one, and factor probes, species, families and viruses. We will also clean some nomenclature for plotting:

```{r Non-retroviral-factor-mutation}

data <- data %>%
  mutate(virus = map_chr(
    virus,
    ~ str_replace(
      string = .,
      pattern = "borna",
      replacement = "Borna"
    )
  )) %>%
  mutate(species = factor(species, levels = sort(unique(species)))) %>%
  mutate(probe_nr = ifelse(probe %in% c("ENV", "GAG", "POL"), probe, "Non-retroviral")) %>%
  mutate(probe_nr = factor(probe_nr, levels = c("ENV", "GAG", "POL", "Non-retroviral"))) %>%
  mutate(family = factor(family, levels = c(sort(unique(
    family[!family %in% c("Bornavirus")]
  )), "Bornavirus"))) %>%
  mutate(virus = factor(virus, levels = c(sort(unique(
    virus
  )))))
```

Let's calculate some distribution parameters:

```{r Distribution}

mean_bit <- mean(data$hsp_bits)
q1_bit <- quantile(data$hsp_bits, 0.25)
median_bit <- quantile(data$hsp_bits, 0.5)
q3_bit <- quantile(data$hsp_bits, 0.75)

q1_bit
median_bit
q3_bit
```

And let's plot it:

```{r Density.Bitscore.plot, fig.height=15, fig.width=25}

density.bitscore.plot <- data %>%
  ggplot() +
  aes(
    x = hsp_bits,
    fill = probe_nr,
    colour = probe_nr,
    weight = hsp_identity,
  ) +
  geom_density(alpha = 0.4) +
  scale_fill_futurama() +
  scale_color_futurama() +
  labs(
    x = "HSP Bitscore",
    y = "Density",
    title = "Distribution of bitscore values",
    subtitle = "Weighted by identity score"
  ) +
  geom_vline(
    xintercept = c(q1_bit, median_bit, q3_bit),
    color = "black",
    linetype = "dashed",
    linewidth = 0.2
  ) +
  annotate("text",
           x = q1_bit - 150,
           y = 1e-4,
           label = "Q1") +
  annotate("text",
           x = median_bit - 150,
           y = 1e-4,
           label = "Q2") +
  annotate("text",
           x = q3_bit - 150,
           y = 1e-4,
           label = "Q3") +
  scale_x_continuous(breaks = seq(0, max(data$hsp_bits), by = 1000)) +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 22L, face = "bold"),
    plot.subtitle = element_text(size = 15L, face = "bold"),
    axis.title.y = element_text(
      size = 15L,
      face = "bold",
      vjust = 5,
      margin = margin(l = 10)
    ),
    axis.title.x = element_text(
      size = 15L,
      face = "bold",
      vjust = -5,
      margin = margin(b = 15)
    ),
    axis.text.x = element_text(size = 12L, face = "bold"),
    axis.text.y = element_text(size = 12L, face = "bold"),
    legend.text = element_text(size = 11L),
    legend.title = element_text(size = 17L)
  )

ggsave("../plots/density-bitscore.png", density.bitscore.plot, height = 15, width = 25)

density.bitscore.plot
```

```{r Raincloud.Bitscore.plot, fig.height=15, fig.width=15}

raincloud.bitscore.plot <- ggplot(data, aes(x = factor(probe_nr, rev(levels(
  probe_nr
))), y = hsp_bits)) +
  geom_boxplot(
    aes(fill = probe_nr),
    position = position_dodge(10),
    outlier.size = 0.001,
    width = 0.1,
    alpha = 0.3
  ) +
  geom_jitter(aes(color = probe_nr),
              size = 0.05,
              width = 0.005,
              alpha = 0.1) +
  stat_halfeye(
    aes(fill = probe_nr),
    alpha = 0.6,
    adjust = 0.5,
    width = 0.8,
    .width = 0,
    justification = -0.4,
    point_interval = NULL
  ) +
  coord_flip() +
  scale_fill_futurama() +
  scale_color_futurama() +
  scale_y_continuous(breaks = seq(0, 10000, by = 1000)) +
  theme_minimal() +
  labs(x = NULL, y = NULL) +
  theme(
    legend.position = "None",
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 16L, face = "bold", vjust = 15),
    plot.margin = margin(r = 20)
  )
                           

ggsave("../plots/raincloud-bitscore.png", raincloud.bitscore.plot, height = 15, width = 20)

raincloud.bitscore.plot
```

Let's take a look at the LTR and HSP sequence lengths:

```{r}

# Let's try generating artificial positions

data <- data %>%
        mutate(ltr_probe_start = ifelse(strand == "+", ltr_start + hsp_sbjct_start, ltr_start + hsp_sbjct_end),
               ltr_probe_end = ifelse(strand == "+", ltr_start + hsp_sbjct_end, ltr_start + hsp_sbjct_start))

head(data$ltr_probe_start)
```

And generate our GenomicRanges:

```{r GenomicRanges}

gr <- GRanges(seqnames = factor(data$accession), 
              ranges = IRanges(
                start = data$ltr_probe_start,
                end = data$ltr_probe_end),
              strand = factor(data$strand))

mcols(gr)$family <- data$family
mcols(gr)$virus <- data$virus
mcols(gr)$bitscore <- data$hsp_bits
mcols(gr)$e_value <- format(data$hsp_evalue, scientific = TRUE)
mcols(gr)$species <- data$species
mcols(gr)$probe <- data$probe
mcols(gr)$accession <- data$accession
mcols(gr)$hsp_start <- data$hsp_sbjct_start
mcols(gr)$hsp_end <- data$hsp_sbjct_end

gr[1234:2348]
```

Let's select a single species to compare against RetroTector. First let's see how many rows does each species contain:

```{r}

species_unique_count <- data %>%
                        group_by(species) %>%
                        summarise(species_count = length(species))

species_unique_count
```

We will go with Hipposideros larvatus. Let's filter the GRanges object for only that species and order the findings by position:

```{r}

gr_hl <- gr[gr$species == 'Hipposideros_larvatus']

un_gr_hl <- sort(unique(gr_hl))

un_gr_hl
```

We can generate GFF3 files from this range:

```{r}

export(un_gr_hl, "F:/Archive/Trabajo/Profesiones/Bioinformatica/PhD/Assignments/enERVate/results/tracks/gr_hl_unique.gff", "GFF3")

```

Let's also try with Molossus nigricans:

```{r}

gr_mn <- gr[gr$species == 'Molossus_nigricans']

un_gr_mn <- sort(unique(gr_mn))

mcols(un_gr_mn)$Name <- paste0(un_gr_mn$probe, ":", ave(seq_along(un_gr_mn$probe), un_gr_mn$probe, FUN = seq_along))

export(un_gr_mn, "F:/Archive/Trabajo/Profesiones/Bioinformatica/PhD/Assignments/enERVate/results/tracks/gr_mn_unique.gff", "GFF3")

un_gr_mn
```

```{r collapsing_ranges}

# Example: reduce overlapping ranges and summarize metadata
reduced.gr.mn <- un_gr_mn %>% group_by(probe, virus, family, species, bitscore) %>% reduce_ranges_directed() %>% arrange(.by_group = start)

reduced.gr.mn$virus <- unfactor(reduced.gr.mn$virus)
reduced.gr.mn$family <- unfactor(reduced.gr.mn$family)
reduced.gr.mn$species <- unfactor(reduced.gr.mn$species)

collapsed.gr.mn <- reduced.gr.mn %>% group_by(probe) %>% reduce(with.revmap = TRUE)

collapsed.gr.mn$probe <- map(collapsed.gr.mn$revmap, function (.) {
                             probe_data <- reduced.gr.mn$probe[.]
                             return(unique(probe_data))})

collapsed.gr.mn$virus <- map(collapsed.gr.mn$revmap, function (.) {
                             virus_data <- reduced.gr.mn$virus[.]
                             return(unique(virus_data))})

collapsed.gr.mn$family <- map(collapsed.gr.mn$revmap, function (.) {
                             family_data <- reduced.gr.mn$family[.]
                             return(unique(family_data))})

collapsed.gr.mn$species <- map(collapsed.gr.mn$revmap, function (.) {
                             species_data <- reduced.gr.mn$species[.]
                             return(unique(species_data))})

collapsed.gr.mn$avg_bitscore <- map(collapsed.gr.mn$revmap, function (.) {
                             bitscore_data <- reduced.gr.mn$bitscore[.]
                             return(c(mean(bitscore_data)))})

collapsed.gr.mn <- collapsed.gr.mn %>%
                                   plyranges::group_by(probe) %>%
                                   plyranges::mutate(probe = as.character(probe),
                                          name = paste0(probe, ':', row_number()
                                   ))

collapsed.gr.mn <- collapsed.gr.mn %>% select(-revmap)

collapsed.gr.mn
```

Let's generate some Sankey diagrams and a bar plot with the distributions. First we generate a grouped variable for plotting:

```{r Grouped.data}

data_counted_probe <- data %>%
                      mutate(species = stringr::str_replace(species, "_", " ")) %>%
                      dplyr::group_by(species, virus, probe_nr, family, abbreviation) %>%
                      dplyr::summarise(count = dplyr::n(), .groups = "keep") %>%
                      dplyr::ungroup()


sum(data_counted_probe[data_counted_probe$probe_nr == "Non-retroviral", ]$count)
```

```{r fig.height=15, fig.width=25}

bar.species_virus.plot <- ggplot(data_counted_probe) +
  aes(
    x = species,
    y = count,
    fill = virus,
    alpha = count
  ) +
  geom_col() +
  scale_fill_manual(values = futurama_unlimited_palette(4, 16)) +
  scale_alpha_continuous(range = c(0.4, 1)) +
  theme_void() +
  labs(fill = "Virus",
       title = "Identified viral integrations in the analysed bat species",
       x = NULL,
       y = "Count",
  ) +
    theme(
        plot.title = element_text(size = 17L, face = "bold", vjust = 3),
        axis.title.x = element_text(size = 12, face = "bold"),     # X-axis title size and bold
        axis.title.y = element_text(size = 12, face = "bold", angle = 90, vjust = 3, margin = margin(l = 10)),     # Y-axis title size and bold
        axis.text.x = element_text(size = 11, face = "bold", angle = 45, vjust = 1, hjust = 1, margin = margin(t = -5)),    # X-axis text size and italic
        axis.text.y = element_text(size = 11, face = "bold"),                     # Y-axis text size (regular)
        strip.text = element_text(size = 12, face = "bold"),       # Facet labels (strip text) size and bold
        legend.title = element_text(size = 13, face = "bold"),     # Legend title size and bold
        legend.text = element_text(size = 10),                     # Legend text size (regular)
        plot.margin = margin(t = 10, l = 30, b = 60),
      ) +
    guides(alpha = "none")

ggsave("../plots/bar_species_virus_plot.png", bar.species_virus.plot, height = 15, width = 25)

bar.species_virus.plot  

```

```{r Sankey.Species.Probe.plot, fig.height=15, fig.width=25}

sankey.species_probe.plot <- ggplot(data_counted_probe,
                                    aes(axis1 = species, axis2 = probe_nr, y = count)) +
  geom_alluvium(aes(fill = species, alpha = count)) +
  geom_stratum(alpha = 0,
               color = "white",
               linewidth = 1.5) +
  geom_text(
    stat = 'stratum',
    aes(label = after_stat(stratum)),
    size = 4L,
    fontface = "bold",
    nudge_x = -0.15,
    hjust = 0
  ) +
  geom_text(
    aes(label = paste0("[", after_stat(count), "]")),
    size = 3.5,
    fontface = "bold",
    nudge_x = 0.15,
    hjust = 0.9,
    stat = "stratum"
  ) +
  theme_void() +
  theme(legend.position = "none") +
  scale_fill_manual(values = futurama_unlimited_palette(12, 100))

ggsave("../plots/sankey-species_probe.png", sankey.species_probe.plot, width = 25, height = 15)

sankey.species_probe.plot
```

```{r Sankey.Species.Family.plot, fig.height=15, fig.width=25}

sankey.species_family.plot <- ggplot(data_counted_probe, aes(axis1 = species, axis2 = family, y = count)) +
  geom_alluvium(aes(fill = species, alpha = count)) +
  geom_stratum(alpha = 0,
               color = "white",
               linewidth = 1.5) +
  geom_text(
    stat = 'stratum',
    aes(label = after_stat(stratum)),
    size = 4L,
    fontface = "bold",
    nudge_x = -0.15,
    hjust = 0
  ) +
  geom_text(
    aes(label = paste0("[", after_stat(count), "]")),
    size = 3.5,
    fontface = "bold",
    nudge_x = 0.15,
    hjust = 0.9,
    stat = "stratum"
  ) +
  theme_void() +
  theme(legend.position = "none") +
  scale_fill_manual(values = futurama_unlimited_palette(12, 100)) +
  labs(title = "Distribution of virus families in bat species", x = NULL, y = NULL) +
  theme(
    plot.title = element_text(
      size = 22L,
      face = "bold",
      vjust = 0,
      hjust = 0.06
    ),
    plot.margin = margin(t = 10),
  )
                      

ggsave("../plots/sankey-species_family.png", sankey.species_family.plot, width = 25, height = 15)

sankey.species_family.plot
```

```{r Sankey.Family.Probe.plot, fig.height=15, fig.width=25}

sankey.family_probe.plot <- ggplot(data_counted_probe,
                                   aes(axis1 = family, axis2 = probe_nr, y = count)) +
  geom_alluvium(aes(fill = family, alpha = count)) +
  geom_stratum(
    aes(fill = family),
    color = "white",
    alpha = 0.1,
    linewidth = 1
  ) +
  geom_text(
    stat = 'stratum',
    aes(label = after_stat(stratum)),
    size = 4L,
    fontface = "bold",
    nudge_x = -0.15,
    hjust = 0
  ) +
  geom_text(
    stat = "stratum",
    aes(label = paste0("[", after_stat(count), "]")),
    size = 3.5,
    fontface = "bold",
    nudge_x = 0.15,
    hjust = 0.9
  ) +
  theme_void() +
  theme(legend.position = "none") +
  scale_fill_futurama() +
  labs(title = "Distribution of probe sequences by virus family", x = NULL, y = NULL) +
  theme(
    plot.title = element_text(
      size = 22L,
      face = "bold",
      vjust = 0,
      hjust = 0.06
    ),
    # Y-axis text size (regular)
    # Legend text size (regular)
    plot.margin = margin(t = 10),
    # legend.position = c(1.3, 0.5)
  )


ggsave("../plots/sankey-probe_family.png", sankey.family_probe.plot, width = 25, height = 15)

sankey.family_probe.plot
```

```{r Sankey.Species.Family.Probe.plot, fig.height=15, fig.width=25}

sankey.species_probe_family.plot <- ggplot(data_counted_probe,
                                           aes(
                                             axis1 = species,
                                             axis2 = family,
                                             axis3 = probe_nr,
                                             y = count
                                           )) +
  geom_alluvium(aes(fill = species, alpha = count)) +
  geom_stratum(alpha = 0,
               color = "white",
               linewidth = 1) +
  geom_text(
    stat = 'stratum',
    aes(label = after_stat(stratum)),
    size = 4L,
    fontface = "bold",
    nudge_x = -0.15,
    hjust = 0
  ) +
  geom_text(
    stat = "stratum",
    aes(label = paste0("[", after_stat(count), "]")),
    size = 3.5,
    fontface = "bold",
    nudge_x = 0.15,
    hjust = 0.9
  ) +
  theme_void() +
  theme(legend.position = "none") +
  scale_fill_manual(values = futurama_unlimited_palette(12, 100)) +
  labs(title = "Global distribution of viral probes and families in bat species", x = NULL, y = NULL) +
  theme(
    plot.title = element_text(
      size = 22L,
      face = "bold",
      vjust = 0,
      hjust = 0.075
    ),
    plot.margin = margin(t = 10),
  )

ggsave("../plots/sankey-species_probe_family.png", sankey.species_probe_family.plot, width = 25, height = 15)

sankey.species_probe_family.plot
```

Let's visualize this last one better with a balloon plot:

```{r Balloon.Virus.Species.Probe.plot, fig.height=15, fig.width=25}



balloon.virus_species.plot <- ggplot(data_counted_probe, aes(x = abbreviation, y = factor(species, levels = rev(unique(
  species
))))) +
  geom_point(aes(
    color = probe_nr,
    shape = probe_nr,
    size = count **  3
  )) +
  scale_color_futurama() +
  facet_wrap( ~ probe_nr) +
  theme_minimal() +
  labs(x = "Virus", y = "Species") +
  labs(title = "Relative abundance of probes per virus within bat species") +
  theme(
    plot.title = element_text(
      size = 22L,
      face = "bold",
      hjust = -0.15
    ),
    axis.title.x = element_text(size = 12, face = "bold"),
    axis.title.y = element_text(size = 12, face = "bold"),
    axis.text.x = element_text(size = 11, face = "bold"),
    axis.text.y = element_text(size = 11, face = "bold"),
    strip.text = element_text(size = 12, face = "bold"),
    legend.title = element_text(size = 12, face = "bold"),
    legend.text = element_text(size = 10)
  ) +
  guides(size = "none")

ggsave("../plots/balloon-virus_species_plot.png", balloon.virus_species.plot, width = 25, height = 15)

balloon.virus_species.plot
```
